File: vdslib/src/vespa/vdslib/state/diskstate.cpp
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "diskstate.h"
#include <boost/lexical_cast.hpp>
#include <vespa/vespalib/text/stringtokenizer.h>
#include <vespa/document/util/stringutil.h>
#include <vespa/vespalib/util/exceptions.h>
#include <vespa/vespalib/stllike/asciistream.h>
#include <vespa/log/log.h>

LOG_SETUP(".vdslib.diskstate");

namespace storage::lib {

DiskState::DiskState()
    : _state(0),
      _description(""),
      _capacity(1.0)
{
    setState(State::UP);
}

DiskState::DiskState(const State& state, const vespalib::stringref & description,
                     double capacity)
    : _state(0),
      _description(description),
      _capacity(1.0)
{
    setState(state);
    setCapacity(capacity);
}

DiskState::DiskState(vespalib::stringref  serialized)
    : _state(&State::UP),
      _description(""),
      _capacity(1.0)
{
    vespalib::StringTokenizer st(serialized, " \t\f\r\n");
    st.removeEmptyTokens();
    for (vespalib::StringTokenizer::Iterator it = st.begin();
         it != st.end(); ++it)
    {
        std::string::size_type index = it->find(':');
        if (index == std::string::npos) {
            throw vespalib::IllegalArgumentException(
                    "Token " + *it + " does not contain ':': " + serialized,
                    VESPA_STRLOC);
        }
        std::string key = it->substr(0, index);
        std::string value = it->substr(index + 1);
        if (key.size() > 0) switch (key[0]) {
            case 's':
                if (key.size() > 1) break;
                setState(State::get(value));
                continue;
            case 'c':
                if (key.size() > 1) break;
                try{
                    setCapacity(boost::lexical_cast<double>(value));
                } catch (...) {
                    throw vespalib::IllegalArgumentException(
                            "Illegal disk capacity '" + value + "'. Capacity "
                            "must be a positive floating point number",
                            VESPA_STRLOC);
                }
                continue;
            case 'm':
                if (key.size() > 1) break;
                _description = document::StringUtil::unescape(value);
                continue;
            default:
                break;
        }
        LOG(debug, "Unknown key %s in diskstate. Ignoring it, assuming it's a "
                   "new feature from a newer version than ourself: %s",
            key.c_str(), vespalib::string(serialized).c_str());
    }

}

void
DiskState::serialize(vespalib::asciistream & out, const vespalib::stringref & prefix,
                     bool includeDescription, bool useOldFormat) const
{
        // Always give node state if not part of a system state
        // to prevent empty serialization
    bool empty = true;
    if (*_state != State::UP || prefix.size() == 0) {
        if (useOldFormat && prefix.size() > 0) {
            out << prefix.substr(0, prefix.size() - 1)
                << ":" << _state->serialize();
        } else {
            out << prefix << "s:" << _state->serialize();
        }
        empty = false;
    }
    if (_capacity != 1.0) {
        if (empty) { empty = false; } else { out << ' '; }
        out << prefix << "c:" << _capacity;
    }
    if (includeDescription && _description.size() > 0) {
        if (empty) { empty = false; } else { out << ' '; }
        out << prefix << "m:"
            << document::StringUtil::escape(_description, ' ');
    }
}


void
DiskState::setState(const State& state)
{
    if (!state.validDiskState()) {
        throw vespalib::IllegalArgumentException(
                "State " + state.toString() + " is not a valid disk state.",
                VESPA_STRLOC);
    }
    _state = &state;
}

void
DiskState::setCapacity(double capacity)
{
    if (capacity < 0) {
        throw vespalib::IllegalArgumentException(
                "Negative capacity makes no sense.", VESPA_STRLOC);
    }
    _capacity = capacity;
}

void
DiskState::print(std::ostream& out, bool verbose,
                 const std::string& indent) const
{
    (void) indent;
    if (verbose) {
        out << "DiskState(" << *_state;
    } else {
        out << _state->serialize();
    }
    if (_capacity != 1.0) {
        out << (verbose ? ", capacity " : ", c ") << _capacity;
    }
    if (_description.size() > 0) {
        out << ": " << _description;
    }
    if (verbose) {
        out << ")";
    }
}

bool
DiskState::operator==(const DiskState& other) const
{
    return (_state == other._state && _capacity == other._capacity);
}

bool
DiskState::operator!=(const DiskState& other) const
{
    return (_state != other._state || _capacity != other._capacity);
}

}

File: vdslib/src/main/java/com/yahoo/vdslib/state/DiskState.java
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.vdslib.state;

import com.yahoo.text.StringUtilities;

import java.util.StringTokenizer;
import java.text.ParseException;

/**
 *
 */
public class DiskState implements Cloneable {
    private State state = State.UP;
    private String description = "";
    private double capacity = 1.0;

    public DiskState() {}
    public DiskState(State s) {
        setState(s);
    }
    public DiskState(State s, String description, double capacity) {
        setState(s); // Set via set methods, so we can have illegal argument checks only one place
        setCapacity(capacity);
        setDescription(description);
    }
    public DiskState clone() {
        try{
            return (DiskState) super.clone();
        } catch (CloneNotSupportedException e) {
            assert(false); // Should not happen
            return null;
        }
    }

    public DiskState(String serialized) throws ParseException {
        StringTokenizer st = new StringTokenizer(serialized, " \t\f\r\n");
        while (st.hasMoreTokens()) {
            String token = st.nextToken();
            int index = token.indexOf(':');
            if (index < 0) {
                throw new ParseException("Token " + token + " does not contain ':': " + serialized, 0);
            }
            String key = token.substring(0, index);
            String value = token.substring(index + 1);
            if (key.length() > 0) switch (key.charAt(0)) {
                case 's':
                    if (key.length() > 1) break;
                    setState(State.get(value));
                    continue;
                case 'c':
                    if (key.length() > 1) break;
                    try{
                        setCapacity(Double.valueOf(value));
                    } catch (Exception e) {
                        throw new ParseException("Illegal disk capacity '" + value + "'. Capacity must be a positive floating point number", 0);
                    }
                    continue;
                case 'm':
                    if (key.length() > 1) break;
                    description = StringUtilities.unescape(value);
                    continue;
                default:
                    break;
            }
            // Ignore unknown tokens
        }
    }

    public String serialize(String prefix, boolean includeDescription) {
        boolean empty = true;
        StringBuilder sb = new StringBuilder();
        if (!state.equals(State.UP) || prefix.length() < 2) {
            sb.append(prefix).append("s:").append(state.serialize());
            empty = false;
        }
        if (Math.abs(capacity - 1.0) > 0.000000001) {
            if (empty) { empty = false; } else { sb.append(' '); }
            sb.append(prefix).append("c:").append(capacity);
        }
        if (includeDescription && description.length() > 0) {
            if (!empty) { sb.append(' '); }
            sb.append(prefix).append("m:").append(StringUtilities.escape(description, ' '));
        }
        return sb.toString();
    }

    public State getState() { return state; }
    public double getCapacity() { return capacity; }
    public String getDescription() { return description; }

    public void setState(State s) {
        if (!s.validDiskState()) {
            throw new IllegalArgumentException("State " + s + " is not a valid disk state.");
        }
        state = s;
    }
    public void setCapacity(double capacity) {
        if (capacity < 0) {
            throw new IllegalArgumentException("Negative capacity makes no sense.");
        }
        this.capacity = capacity;
    }
    public void setDescription(String desc) { description = desc; }

    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("DiskState(").append(state.serialize());
        if (Math.abs(capacity - 1.0) > 0.00000001) {
            sb.append(", capacity ").append(capacity);
        }
        if (description.length() > 0) {
            sb.append(": ").append(description);
        }
        sb.append(")");
        return sb.toString();
    }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof DiskState)) { return false; }
        DiskState other = (DiskState) o;
        if (state.equals(other.state)
            && Math.abs(capacity - other.capacity) < 0.00000001)
        {
            return true;
        }
        return false;
    }

    @Override
    public int hashCode() {
        // NOTE: capacity cannot be part of the hashCode
        return state.hashCode();
    }
}
